File: chapter15/src/main/java/com/apress/prospring6/fifteen/controllers/SingerController.java
package com.apress.prospring6.fifteen.controllers;

import com.apress.prospring6.fifteen.entities.Singer;
import com.apress.prospring6.fifteen.repos.SingerRepo;
import jakarta.validation.Valid;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.dao.DataIntegrityViolationException;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;

@RestController
@RequestMapping(path = "singer")
public class SingerController {

    final Logger logger = LoggerFactory.getLogger(SingerController.class);

    private final SingerRepo singerRepo;

    public SingerController(SingerRepo singerRepo) {
        this.singerRepo = singerRepo;
    }

    @ResponseStatus(HttpStatus.OK)
    @GetMapping("/")
    public List<Singer> all() {
        return singerRepo.findAll();
    }

    @GetMapping("/{id}")
    public ResponseEntity<Singer> findSingerById(@PathVariable("id") Long id) {
        return singerRepo.findById(id)
                .map(ResponseEntity::ok)
                .orElseGet(() -> ResponseEntity.status(HttpStatus.NOT_FOUND).body(new Singer()));
    }

    @PostMapping("/")
    public ResponseEntity<?> create(@RequestBody @Valid Singer singer) {
        logger.info("Creating singer: " + singer);
        try {
            var saved = singerRepo.save(singer);
            logger.info("Singer created successfully with info: " + saved);
            return new ResponseEntity<>(saved, HttpStatus.CREATED);
        } catch (DataIntegrityViolationException dive) {
            logger.debug("Could not create singer.", dive);
            return ResponseEntity.badRequest().body(dive.getMessage());
        }
    }

    @ResponseStatus(HttpStatus.OK)
    @PutMapping("/{id}")
    public void update(@RequestBody Singer singer, @PathVariable("id") Long id) {
        logger.info("Update singer: {}", singer);
        var fromDb = singerRepo.findById(id).orElseThrow(() -> new IllegalArgumentException("Singer does not exists"));

        fromDb.setFirstName(singer.getFirstName());
        fromDb.setLastName(singer.getLastName());
        fromDb.setBirthDate(singer.getBirthDate());

        singerRepo.save(fromDb);

        logger.info("Singer updated successfully with info: " + fromDb);
    }

    @ResponseStatus(HttpStatus.NO_CONTENT)
    @DeleteMapping("/{id}")
    public void delete(@PathVariable("id") Long id) {
        logger.info("Deleting singer with id: " + id);
        singerRepo.deleteById(id);
        logger.info("Singer deleted successfully");
    }
}

File: chapter16-kotlin/src/main/kotlin/com/apress/prospring6/sixteen/kotlin/boot/SingerController.kt
package com.apress.prospring6.sixteen.kotlin.boot

import jakarta.validation.Valid
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/singer")
class SingerController (private val service: SingerService) {

    @GetMapping
    fun all(): List<Singer> {
        return service.findAll()
    }

    @GetMapping("/{id}")
    fun findSingerById(@PathVariable id: Long): Singer? {
        return service.findById(id)
    }

    @PostMapping
    fun create(@RequestBody @Valid singer: Singer): Singer? {
        return service.save(singer)
    }

    @PutMapping("/{id}")
    fun update(@RequestBody @Valid singer: Singer, @PathVariable id: Long): Singer? {
        return service.update(id, singer)
    }

    @DeleteMapping("/{id}")
    fun delete(@PathVariable id: Long) {
        service.delete(id)
    }
}